Panic on zero length array in exportArray

### Describe the bug, including details regarding any error messages, version, and platform.

I have seen some panics at line 410 in  https://github.com/apache/arrow-go/blob/56b794f52a9bfb45fb54a4e070002a69b38fb018/arrow/cdata/cdata_exports.go#L403-L410

This is when the array `childPtrs` is of size 0.

Its hard to replicate, so at present my question is simply if zero length arrays should be checked for here? I assume this would be a record with 0 fields?

The panics were with a July commit of release candidate for v18 (linux x86) but wanted to check first if this is a code issue while I try to replicate.

```golang
panic: runtime error: index out of range [0] with length 0

goroutine 17 [running, locked to thread]:
github.com/apache/arrow/go/v18/arrow/cdata.exportArray({0x5564f20, 0xc001d91000}, 0x7fac48074650, 0x0)
        /home/clive/go/pkg/mod/github.com/apache/arrow/go/v18@v18.0.0-20240722165357-ec58e4de9e08/arrow/cdata/cdata_exports.go:406 +0x10a5
github.com/apache/arrow/go/v18/arrow/cdata.ExportArrowRecordBatch({0x55609b8, 0xc001509a70}, 0x7fac48074650, 0x0)
        /home/clive/go/pkg/mod/github.com/apache/arrow/go/v18@v18.0.0-20240722165357-ec58e4de9e08/arrow/cdata/interface.go:250 +0x436
github.com/apache/arrow/go/v18/arrow/cdata.cRecordReader.next({{0x554da68, 0xc001fbe140}, 0x0}, 0x7fac48074650)
        /home/clive/go/pkg/mod/github.com/apache/arrow/go/v18@v18.0.0-20240722165357-ec58e4de9e08/arrow/cdata/cdata_exports.go:449 +0x71
github.com/apache/arrow/go/v18/arrow/cdata.streamGetNext(0x7faccc00c890, 0x7fac48074650)
        /home/clive/go/pkg/mod/github.com/apache/arrow/go/v18@v18.0.0-20240722165357-ec58e4de9e08/arrow/cdata/exports.go:129 +0xbd

```

### Component(s)

Release

--- a/arrow/cdata/cdata_exports.go
+++ b/arrow/cdata/cdata_exports.go
@@ -401,6 +401,9 @@ func exportArray(arr arrow.Array, out *CArrowArray, outSchema *CArrowSchema) {
 		out.children = (**CArrowArray)(unsafe.Pointer(&childPtrs[0]))
 	case *array.Struct:
 		out.n_children = C.int64_t(arr.NumField())
+		if arr.NumField() == 0 {
+			return
+		}
 		childPtrs := allocateArrowArrayPtrArr(arr.NumField())
 		children := allocateArrowArrayArr(arr.NumField())
 		for i := 0; i < arr.NumField(); i++ {

--- a/arrow/cdata/cdata_exports.go
+++ b/arrow/cdata/cdata_exports.go
@@ -424,6 +424,10 @@ func exportArray(arr arrow.Array, out *CArrowArray, outSchema *CArrowSchema) {
 		exportArray(arr.Dictionary(), out.dictionary, nil)
 	case array.Union:
 		out.n_children = C.int64_t(arr.NumFields())
+		if arr.NumFields() == 0 {
+			return
+		}
+
 		childPtrs := allocateArrowArrayPtrArr(arr.NumFields())
 		children := allocateArrowArrayArr(arr.NumFields())
 		for i := 0; i < arr.NumFields(); i++ {

--- a/arrow/cdata/cdata_test.go
+++ b/arrow/cdata/cdata_test.go
@@ -598,6 +598,14 @@ func createTestStructArr() arrow.Array {
 	return bld.NewArray()
 }
 
+func createTestEmptyStructArr() arrow.Array {
+	bld := array.NewStructBuilder(memory.DefaultAllocator, arrow.StructOf())
+	defer bld.Release()
+
+	bld.AppendNull()
+	return bld.NewArray()
+}
+
 func createTestRunEndsArr() arrow.Array {
 	bld := array.NewRunEndEncodedBuilder(memory.DefaultAllocator,
 		arrow.PrimitiveTypes.Int32, arrow.PrimitiveTypes.Int8)
@@ -687,6 +695,7 @@ func TestNestedArrays(t *testing.T) {
 		{"sparse union", createTestSparseUnion},
 		{"dense union", createTestDenseUnion},
 		{"run-end encoded", createTestRunEndsArr},
+		{"empty struct", createTestEmptyStructArr},
 	}
 
 	for _, tt := range tests {

--- a/arrow/cdata/cdata_test.go
+++ b/arrow/cdata/cdata_test.go
@@ -606,6 +606,20 @@ func createTestEmptyStructArr() arrow.Array {
 	return bld.NewArray()
 }
 
+func createTestEmptyDenseUnionArr() arrow.Array {
+	bld := array.NewEmptyDenseUnionBuilder(memory.DefaultAllocator)
+	defer bld.Release()
+
+	return bld.NewArray()
+}
+
+func createTestEmptySparseUnionArr() arrow.Array {
+	bld := array.NewEmptySparseUnionBuilder(memory.DefaultAllocator)
+	defer bld.Release()
+
+	return bld.NewArray()
+}
+
 func createTestRunEndsArr() arrow.Array {
 	bld := array.NewRunEndEncodedBuilder(memory.DefaultAllocator,
 		arrow.PrimitiveTypes.Int32, arrow.PrimitiveTypes.Int8)
@@ -696,6 +710,8 @@ func TestNestedArrays(t *testing.T) {
 		{"dense union", createTestDenseUnion},
 		{"run-end encoded", createTestRunEndsArr},
 		{"empty struct", createTestEmptyStructArr},
+		{"empty dense union", createTestEmptyDenseUnionArr},
+		{"empty sparse union", createTestEmptySparseUnionArr},
 	}
 
 	for _, tt := range tests {

--- a/arrow/cdata/cdata_test_framework.go
+++ b/arrow/cdata/cdata_test_framework.go
@@ -309,6 +309,9 @@ func createCArr(arr arrow.Array, alloc *mallocator.Mallocator) *CArrowArray {
 		children = (**CArrowArray)(unsafe.Pointer(&clist[0]))
 		nchildren += 1
 	case *array.Struct:
+		if arr.NumField() == 0 {
+			break
+		}
 		clist := allocateChildrenPtrArr(alloc, arr.NumField())
 		for i := 0; i < arr.NumField(); i++ {
 			clist[i] = createCArr(arr.Field(i), alloc)

--- a/arrow/cdata/cdata_test_framework.go
+++ b/arrow/cdata/cdata_test_framework.go
@@ -325,6 +325,9 @@ func createCArr(arr arrow.Array, alloc *mallocator.Mallocator) *CArrowArray {
 		children = (**CArrowArray)(unsafe.Pointer(&clist[0]))
 		nchildren += 2
 	case array.Union:
+		if arr.NumFields() == 0 {
+			break
+		}
 		clist := allocateChildrenPtrArr(alloc, arr.NumFields())
 		for i := 0; i < arr.NumFields(); i++ {
 			clist[i] = createCArr(arr.Field(i), alloc)
@@ -359,7 +362,7 @@ func createCArr(arr arrow.Array, alloc *mallocator.Mallocator) *CArrowArray {
 	tr.bufs = make([][]byte, 0, nbuffers)
 	cbufs := allocateBufferMallocatorPtrArr(alloc, nbuffers)
 	for i, b := range buffers[bufOffset:] {
-		if b != nil {
+		if b != nil && b.Len() > 0 {
 			raw := alloc.Allocate(b.Len())
 			copy(raw, b.Bytes())
 			tr.bufs = append(tr.bufs, raw)

